sd-boot: add support for a sysfail entry

igoropaniuk opened the pull request:
Add support for a sysfail boot entry. Sysfail boot entries can be used for optional tweaking the automatic selection order in case a failure state of the system in some form is detected (boot firmware failure etc).

The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry to use. It can be set using bootctl:
```
$ bootctl set-sysfail sysfail.conf
```

The `LoaderEntrySysFail` EFI variable would be unset automatically during next boot by `systemd-boot-clear-sysfail.service` if no system failure occured, otherwise it would be kept as it is and a system
failure reason will be saved to `LoaderSysFailReason` EFI variable.

`sysfail_check()` expected to be extented to support possibleconditions when we should boot sysfail("recovery") boot entry.

Also add support for using a sysfail boot entry in case of UEFI firmware capsule update failure [1]. The status of a firmware update is obtained from the EFI System Resource Table (ESRT), which provides an optional mechanism for identifying device and system firmware resources for the purposes of targeting firmware updates to those resources.
    
Current implementation uses the value of LastAttemptStatus field from ESRT, which describes the result of the last firmware update attempt for the firmware resource entry. The field is updated each time an
`UpdateCapsule()` is attempted for an ESRT entry and is preserved across reboots (non-volatile).
    
This can be be used in setups with support for A/B OTA updates, where the boot firmware and Linux/RootFS might be updated synchronously.
    
The check is activated by adding "sysfail-firmware-upd" to loader.conf
    
[1] https://uefi.org/specs/UEFI/2.10/23_Firmware_Update_and_Reporting.html

bluca commented:
We already have boot success/failure counting which can do A/B/C and more, why is this additional mechanism needed?

igoropaniuk commented:
@bluca Thanks for taking a look into this
> We already have boot success/failure counting which can do A/B/C and more, why is this additional mechanism needed?

Well, the existing Boot Assesment mechanism will iterate over all available boot entries (decreasing a boot counter etc), until boot is blessed. This particular change provides an additional mechanism, which permits to jump directly to the "recovery" entry if some condition is fulfilled, avoiding "probing" all available boot entries.

I would say that they can complement each other and function in parallel

bluca commented:
But you can simply name the available entries accordingly to do the same? WIthout the need for any additional EFIVAR

igoropaniuk commented:
> But you can simply name the available entries accordingly to do the same? WIthout the need for any additional EFIVAR
Right, but even with only two entries available (one of which is a "recovery" entry), systemd-boot will jump to "recovery"  only after MAX_TRIES fails of previous entry.
My point is that why not to have a complementray mechanism to jump to the correct one ASAP without engaging boot counting

bluca commented:
Just set the max tries to 1?

mbolivar commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+#include "recovery.h"
+#include "util.h"
+
+#define FALLBACK_BOOT_ENTRY L"ostree-1-lmp.conf\0"
```
Is this needed?

poettering commented:
hmm, when is this esrt field reset? never? on the next attempted bios update? how precisely does that work?

I am a bit concerned about introducing two orthogonal fallback mechanisms, i.e. the boot counter based one, and this one that operates quite differently?

I guess I am fine if we change the default item selection logic in case we detect the system is in a unclean state. That makes a ton of sense to me, but I'd probably not call that "fallback". But maybe "recovery" or so?

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1777,11 +1789,29 @@ static size_t config_find_entry(Config *config, const char16_t *pattern) {
+
+        if (config->fallback_firmware_failure) {
+                return recovery_check_firmware_failure();
+        }
```
no {} round single line if blocks

poettering commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+        EFI_SYSTEM_RESOURCE_ENTRY *esrt_entry = NULL;
+
+        esrt_table = find_configuration_table(MAKE_GUID_PTR(EFI_SYSTEM_RESOURCE_TABLE));
+        if (esrt_table) {
```
we prefer early-exit over deep-nesting. i.e. change this line to `if (!esrt_table)\n return false;` so that the rest of the function doesn't have to be indented

poettering commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+
+bool recovery_check_firmware_failure(void) {
+        EFI_SYSTEM_RESOURCE_TABLE *esrt_table = NULL;
+        EFI_SYSTEM_RESOURCE_ENTRY *esrt_entry = NULL;
```
I'd make these const

poettering commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+
+        esrt_table = find_configuration_table(MAKE_GUID_PTR(EFI_SYSTEM_RESOURCE_TABLE));
+        if (esrt_table) {
+                esrt_entry = (void *)(esrt_table + 1);
```
if you do +1 here, this always suggests to the reader that we are talking of an array here, but we aren't really, you just skip the header. hence i'd prefer if you'd write this like this:

```c
esrt_entry = (const EFI_SYSTEM_RESOURCE_ENTRY*) ((const uint8_t*) esrt_table + sizeof(EFI_SYSTEM_RESOURCE_TABLE));
```

even if this is longer it's immediately clear what this is

poettering commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+                for (uint32_t i = 0; i < esrt_table->FwResourceCount; i++, esrt_entry++) {
+                        if (esrt_entry->FwType == ESRT_FW_TYPE_SYSTEMFIRMWARE) {
+                                return (esrt_entry->LastAttemptStatus != LAST_ATTEMPT_STATUS_SUCCESS);
+                        }
```
no {} around single line if block.

the () on the return line is redundant

poettering commented on src/boot/efi/recovery.c:
```diff
@@ -0,0 +1,23 @@
+        }
+
+        return false;
+}
```
trailing newline is missing

poettering commented on src/boot/efi/recovery.h:
```diff
@@ -0,0 +1,7 @@
+#include <efi.h>
+#include "efivars-fundamental.h"
+
+bool recovery_check_firmware_failure(void);
```
trailing newline is missing

poettering commented on src/boot/efi/recovery.h:
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <efi.h>
+#include "efivars-fundamental.h"
```
this include seems unnecessary in the header?

poettering commented:
So let me get this straight, after reading the code in detail: the concept is not really about fallback of bad menu items at all, the first paragraph in the commit is hence quite misleading. 

Instead, this is about optional tweaking the automatic selection order in case a failure state of the system in some form is detected. Did I get this right?

I think expressed like this the concept does make a lot of sense to me. But it really shouldn't be called "fallback" in that case, but more "recovery" or "sysfail" or something like that.

I am also concerned about how the efi var for this is controlled. What's your model there? would a firmware update tool set that efi var before it initiates the firmware update, and afterwards it would unset it? how long would you expect the efi var to remain set? just one boot?

if so, shouldn't we ship a unit file by default that unsets the efi var if it exists if we managed to successfully boot?

poettering commented:
> I've been also thinking about extending `bless-boot` tool instead of having an additional unit file, let me know if this option works for you. Thanks!

I'd keep that separate. bless-boot is potentially delayed for a long time, and this flag seems like something we can pretty quickly reset.

poettering commented:
I think we should clearly communicate to userspace why we entered sysfail state, i.e. define a volatile efi var for that, and since for now we only have one reason for this, write a constant value into it "esrt-firmware-update-error" or so. something like that. And we the tool that unsets the nv var should probably read the volatile var and log about it.

so i am onboard with the concept i guess. but please change the name from "fallback" to something else as discussed.

igoropaniuk commented:
@poettering I've finally pushed v2, sorry for a long delay.

I've addressed all comments (please let me know if I missed something), including:
1. Using different naming for this feature: `sysfail entry` 
2. Worflow changed a bit:
    The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry to use. It can be set using bootctl:
    $ bootctl set-sysfail sysfail.conf
    
    The `LoaderEntrySysFail` EFI variable would be unset automatically during next boot by `systemd-boot-clear sysfail.service` if no system failure occured, otherwise it would be kept as it is and a system failure reason will be saved to `LoaderSysFailReason` EFI variable.

igoropaniuk commented on units/systemd-boot-clear-sysfail.service:
```diff
@@ -0,0 +1,28 @@
+ExecStart=bootctl set-sysfail ""
+
+[Install]
+WantedBy=sysinit.target
```
Will add trailing newline

igoropaniuk commented on units/systemd-boot-clear-sysfail.service:
```diff
@@ -0,0 +1,28 @@
+Before=sysinit.target
+
+# If LoaderSysFailReason is set we should not clear SysFail entry
+ConditionPathExists=!/sys/firmware/efi/efivars/LoaderSysFailReason-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f
```
@poettering Having this conditional prevents `LoaderEntrySysFail` to be cleared in case system failure occurs (and error is reported in `LoaderSysFailReason`). However this leads to failed integration tests:
```
[FAILED] Failed to start systemd-boot-clear-sysfail.service.
```
I was thinking about adding a simple `if` bash condition instead:
```
ExecStart=sh -c "if [ ! -f $/sys/firmware/efi/efivars/LoaderSysFailReason-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f ]; then exit 0; else bootctl set-sysfail ""; fi"
```

Any advise how that can be addressed in a better way?

igoropaniuk replied:
Turned out that the root cause of the integration test failure was different

igoropaniuk commented:
@poettering gentle ping, let me know if there is anything else for me to do in the context of this PR

yuwata commented on docs/BOOT_LOADER_INTERFACE.md:
```diff
@@ -58,6 +58,17 @@ variables. All EFI variables use the vendor UUID
+  to use. Sysfail boot entries can be used optional tweaking the automatic
+  selection order in case a failure state of the system in some form is detected.
+  If system failure occurs and `LoaderEntrySysFail` is set, systemd will
+  use this boot entry and actual sysfail reason to `LoaderEntrySysFail` EFI
```
LoaderEntrySysFail -> LoaderSysFailReason ?

igoropaniuk replied:
thanks, fixed!

igoropaniuk commented:
I see two failed checks: `noble/arm64 ` and `noble/ppc64el`, but seems that they are unrelated

yuwata commented on docs/BOOT_LOADER_INTERFACE.md:
```diff
@@ -58,6 +58,17 @@ variables. All EFI variables use the vendor UUID
   to use. It contains a NUL-terminated boot loader entry identifier.
 
+* The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry
+  to use. Sysfail boot entries can be used optional tweaking the automatic
```
I am not native, but maybe
`can be used optional tweaking` -> `can be used for optionally tweaking`
??

yuwata commented on docs/BOOT_LOADER_INTERFACE.md:
```diff
@@ -58,6 +58,17 @@ variables. All EFI variables use the vendor UUID
 
+* The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry
+  to use. Sysfail boot entries can be used optional tweaking the automatic
+  selection order in case a failure state of the system in some form is detected.
```
It sounds ambiguous. First of all, failure here means something in boot sequence?
Please declare more explicitly. And, may be better to add some example failure cases.

igoropaniuk replied:
I've completely rephrased the whole description

yuwata commented on docs/BOOT_LOADER_INTERFACE.md:
```diff
@@ -58,6 +58,17 @@ variables. All EFI variables use the vendor UUID
+* The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry
+  to use. Sysfail boot entries can be used optional tweaking the automatic
+  selection order in case a failure state of the system in some form is detected.
+  If system failure occurs and `LoaderEntrySysFail` is set, systemd will
```
Maybe, `systemd will use` -> `systemd-boot will use` ??

yuwata commented on docs/BOOT_LOADER_INTERFACE.md:
```diff
@@ -58,6 +58,17 @@ variables. All EFI variables use the vendor UUID
 * The EFI variable `LoaderEntryDefault` contains the default boot loader entry
   to use. It contains a NUL-terminated boot loader entry identifier.
 
+* The EFI variable `LoaderEntrySysFail` contains the sysfail boot loader entry
```
`the sysfail boot loader entry` sounds ambiguous, as `sysfail` is not declared at this point...

yuwata commented on src/boot/boot.c:
```diff
@@ -531,6 +533,8 @@ static void print_status(Config *config, char16_t *loaded_image_path) {
         if (config->entry_saved)
                 printf("           saved entry: %ls\n", config->entry_saved);
+        if (config->entry_sysfail)
+                printf("    default (sysfail): %ls\n", config->entry_sysfail);
```
Please align with `:`, that is, please insert one more space before `default`.

yuwata commented on src/boot/meson.build:
```diff
@@ -295,6 +295,7 @@ libefi_sources = files(
         'measure.c',
         'part-discovery.c',
         'pe.c',
+        'sysfail.c',
```
Please sort.

yuwata commented on src/boot/sysfail.h:
```diff
@@ -0,0 +1,14 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <efivars-fundamental.h>
```
Please use `"..."` rather than `<...>` for headers we provide.

yuwata commented on src/boot/sysfail.h:
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <efivars-fundamental.h>
+#include <uchar.h>
```
Please drop this. See comments in efi.h .

yuwata commented on units/systemd-boot-clear-sysfail.service:
```diff
@@ -0,0 +1,30 @@
+Documentation=man:systemd-boot-sysfail.service(8)
+
+DefaultDependencies=no
+After=local-fs.target systemd-random-seed.service
```
Could you add a comment here why this unit is ordered after them?

yuwata commented on src/bootctl/bootctl.c:
```diff
@@ -629,6 +629,7 @@ static int bootctl_main(int argc, char *argv[]) {
                 { "set-oneshot",         2,        2,        0,            verb_set_efivar          },
                 { "set-timeout",         2,        2,        0,            verb_set_efivar          },
                 { "set-timeout-oneshot", 2,        2,        0,            verb_set_efivar          },
+                { "set-sysfail",         2,        2,        0,            verb_set_efivar          },
```
Please update help message and man page.

yuwata commented on docs/BOOT.md:
```diff
@@ -102,6 +102,7 @@ Some EFI variables control the loader or exported the loaders state to the start
 | EFI Variables |
 |---------------|------------------------|-------------------------------|
 | LoaderEntryDefault | entry identifier to select as default at bootup  | non-volatile |
+| LoaderEntrySysFail | sysfail entry identifier | non-volatile |
```
Maybe LoaderSysFailReason should be mentioned in the table or somewhere?

yuwata commented on src/boot/boot.c:
```diff
@@ -1787,7 +1794,9 @@ static bool config_sysfail_occured(Config *config) {
         assert(config);
 
-        sysfail_type = sysfail_check();
+        SysFailConfig sysfail_config = { .check_firmware_update = config->sysfail_firmware_upd};
```
Please wrap the line like the other places:
```c
SysFailConfig sysfail_config = { 
        .check_firmware_update = config->sysfail_firmware_upd,
};
```

igoropaniuk replied:
fixed

yuwata commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
-SysFailType sysfail_check(void) {
+
+static bool check_firmware_update_failure(void) {
+        const EFI_SYSTEM_RESOURCE_TABLE *esrt_table = NULL;
+        const EFI_SYSTEM_RESOURCE_ENTRY *esrt_entry = NULL;
```
Unnecessary initialization. Please drop `= NULL`.

yuwata commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
+        if (!esrt_table)
+                return false;
+
+        esrt_entry = (const EFI_SYSTEM_RESOURCE_ENTRY*) ((const uint8_t*) esrt_table +
+                        sizeof(EFI_SYSTEM_RESOURCE_TABLE));
```
I do not think it is not necessary to break line.

yuwata commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
+        esrt_entry = (const EFI_SYSTEM_RESOURCE_ENTRY*) ((const uint8_t*) esrt_table +
+                        sizeof(EFI_SYSTEM_RESOURCE_TABLE));
+
+        for (uint32_t i = 0; i < esrt_table->FwResourceCount; i++, esrt_entry++)
+                        if (esrt_entry->FwType == ESRT_FW_TYPE_SYSTEMFIRMWARE)
+                                return esrt_entry->LastAttemptStatus != LAST_ATTEMPT_STATUS_SUCCESS;
```
Please reindent, and use 8ch indentation.

yuwata commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
+        esrt_entry = (const EFI_SYSTEM_RESOURCE_ENTRY*) ((const uint8_t*) esrt_table +
+                        sizeof(EFI_SYSTEM_RESOURCE_TABLE));
+
+        for (uint32_t i = 0; i < esrt_table->FwResourceCount; i++, esrt_entry++)
```
Please use `FOREACH_ARRAY()` macro.

yuwata commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
 
-SysFailType sysfail_check(void) {
+
+static bool check_firmware_update_failure(void) {
```
boolean function named `check_xyz_failure()` sounds confusing in the result. Without reading the function definition, we cannot know the meaning of `true/false`.
Please rename e.g. `find_firmware_update_failure()` or `firmware_update_is_failed()`.

yuwata commented on src/boot/sysfail.h:
```diff
@@ -6,9 +6,14 @@
 
 typedef enum SysFailType {
         SYSFAIL_OK = 0x0,
```
The naming `SYSFAIL_OK` sounds spurious. `FAIL_OK` means succeeds? Or successfully fail??

igoropaniuk replied:
Do you have any suggestions how it might be called in a better way?

yuwata commented on src/boot/boot.c:
```diff
@@ -1776,11 +1782,35 @@ static size_t config_find_entry(Config *config, const char16_t *pattern) {
+        assert(config);
+
+        sysfail_type = sysfail_check();
+        if (sysfail_type != SYSFAIL_OK) {
```
Please invert the condition and return earlier.

igoropaniuk replied:
fixed

igoropaniuk commented:
@yuwata, thank you very much for your review and valuable comments/suggestions.

I've addressed almost all comments and left additional questions

yuwata reviewed (changes requested):
I am not a fan of the naming `SysFail`. First of all, we usually do not abbreviate setting name. And SysFail (or SystemFailure) is quite ambiguous in
- what kind of failures??
- what will be done by the entry??

How about `LoaderEntryFallback` or so? That's still ambiguous, but I think it is at least better than sysfail. Of course, if you have any better name, that's OK.

igoropaniuk commented:
@yuwata 

> How about LoaderEntryFallback or so? That's still ambiguous, but I think it is at least better than sysfail. Of course, if you have any better name, that's OK.

Actually, it has been called that from the very beginning. However, @poettering suggested ([his original comment](https://github.com/systemd/systemd/pull/34856#issuecomment-2431413327)) that 'fallback' might be misleading:

> But it really shouldn't be called "fallback" in that case, but more "recovery" or "sysfail" or something like that.

poettering commented on src/boot/sysfail.h:
```diff
@@ -0,0 +1,13 @@
+#include "efivars-fundamental.h"
+
+typedef enum SysFailType {
+        SYSFAIL_NO_FAILURE = 0x0,
```
the assignment is redundant

poettering commented on src/boot/sysfail.h:
```diff
@@ -0,0 +1,13 @@
+typedef enum SysFailType {
+        SYSFAIL_NO_FAILURE = 0x0,
+        _SYSFAIL_MAX,
+        _SYSFAIL_INVALID = -EINVAL,
```
EINVAL is a userspace thing, shouldn't be in src/boot/sysfail.h, i.e. a boot lodaer header

poettering commented on units/systemd-boot-clear-sysfail.service:
```diff
@@ -0,0 +1,30 @@
+#  (at your option) any later version.
+
+[Unit]
+Description=Clear SysFail entry if the boot is successful
```
We Usually Use Title Case For Description= Fields

poettering commented on src/shared/bootspec.c:
```diff
@@ -1433,6 +1433,12 @@ static int boot_load_efi_entry_pointers(BootConfig *config, bool skip_efivars) {
         if (r < 0 && !IN_SET(r, -ENOENT, -ENODATA))
                 log_warning_errno(r, "Failed to read EFI variable \"LoaderEntrySelected\", ignoring: %m");
 
+        r = efi_get_variable_string(EFI_LOADER_VARIABLE_STR("LoaderEntrySysFail"), &config->entry_sysfail);
```
the field needs to be freed too eventually

poettering commented on src/boot/boot.c:
```diff
@@ -1776,11 +1782,34 @@ static size_t config_find_entry(Config *config, const char16_t *pattern) {
+                return false;
+
+        /* Store reason string in LoaderSysFailReason EFI variable */
+        const char16_t *reason_str = sysfail_get_error_str(sysfail_type);
```
hmm, please show this field also in the status output of bootctl, if set. i.e. in bootctl-status.c

poettering commented on src/bootctl/bootctl.c:
```diff
@@ -237,6 +237,7 @@ static int help(int argc, char *argv[], void *userdata) {
                "\n%3$sBoot Loader Interface Commands:%4$s\n"
                "  set-default ID       Set default boot loader entry\n"
                "  set-oneshot ID       Set default boot loader entry, for next boot only\n"
+               "  set-sysfail ID       Set boot loader entry to be used in case of a sys failure\n"
```
in the English explanatory string please don't abbreviate "system" unnecessarily, i.e. on the right hand side please write out "system failure"

poettering commented on units/systemd-boot-clear-sysfail.service:
```diff
@@ -0,0 +1,30 @@
+
+[Unit]
+Description=Clear SysFail entry if the boot is successful
+Documentation=man:systemd-boot-sysfail.service(8)
```
actually needs a man page

poettering commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
+        case SYSFAIL_NO_FAILURE:
+                return NULL;
+        case SYSFAIL_FIRMWARE_UPDATE:
+                return u"FirmwareUpdateFailed";
```
in systemd we usually write enum values in lowercase with dashes. hence maybe return "firmware-update-failed" here.

poettering replied:
also, please make sure that the string and the C name are the same (except for case and separators, and a prefix). Hence maybe SYSFAIL_FIRMWARE_UPDATE_FAILURE and "firmware-updare-failure".

poettering commented on src/boot/efi.h:
```diff
@@ -426,6 +436,22 @@ typedef struct {
+        uint32_t FwResourceCount;
+        uint32_t FwResourceCountMax;
+        uint64_t FwResourceVersion;
+} EFI_SYSTEM_RESOURCE_TABLE;
```
maybe end this in `EFI_SYSTEM_RESOURCE_ENTRY Entries[]`?

poettering commented on src/boot/efi.h:
```diff
@@ -126,6 +126,16 @@ typedef uint64_t EFI_PHYSICAL_ADDRESS;
+#define ESRT_FW_TYPE_DEVICEFIRMWARE 0x00000002
+#define ESRT_FW_TYPE_UEFIDRIVER     0x00000003
+
+#define LAST_ATTEMPT_STATUS_SUCCESS 0x00000000
```
please suffix these with U, otherwise you'll always compare signed with unsigned integers, wich is always a bit icky

poettering commented on src/boot/sysfail.c:
```diff
@@ -3,10 +3,37 @@
+        if (!esrt_table)
+                return false;
+
+        esrt_entries = (const EFI_SYSTEM_RESOURCE_ENTRY*)((const uint8_t*)esrt_table + sizeof(EFI_SYSTEM_RESOURCE_TABLE));
```
this would be much nicer if you did the Entries[] thing i suggested above

poettering commented on src/boot/boot.c:
```diff
@@ -1787,7 +1794,11 @@ static bool config_sysfail_occured(Config *config) {
-        sysfail_type = sysfail_check();
+        SysFailConfig sysfail_config = {
+                .check_firmware_update = config->sysfail_firmware_update
+        };
```
why make this configurable? seems unnecessary to me? i mean the SysFail entry is not going to be set by default, hence we aren't going to do anything in the default case anyway, hence seems safe to just enable this

poettering commented:
oh, could you add the userspace parts in a separate commit from the uefi parts?

poettering commented:
> > How about LoaderEntryFallback or so? That's still ambiguous, but I think it is at least better than sysfail. Of course, if you have any better name, that's OK.
> 
> Actually, it has been called that way from the very beginning. However, @poettering suggested ([his original comment](https://github.com/systemd/systemd/pull/34856#issuecomment-2431413327)) that 'fallback' might be misleading:
> 
> > But it really shouldn't be called "fallback" in that case, but more "recovery" or "sysfail" or something like that.

so i don't like "fallback", for the aforementioned reasons. But I am not glued to "sysfail". If somebody can make a better suggestion I am fine with it. @yuwata?

igoropaniuk commented:
@poettering @yuwata All comments have been addressed.

We still have an open question regarding the naming. Please let me know if you come up with something more suitable than "sysfail"

poettering commented on src/boot/boot.c:
```diff
@@ -1673,11 +1680,36 @@ static size_t config_find_entry(Config *config, const char16_t *pattern) {
         assert(config);
 
+        i = config_find_entry(config, config->entry_sysfail);
+        if (config->sysfail_occured && i != IDX_INVALID) {
```
please check sysfail_occured first, and then go looking for a suitable entry. seems surprising to do it the other way round

igoropaniuk replied:
@poettering sorry, missed that, fixed. thanks!

poettering commented on src/bootctl/bootctl-status.c:
```diff
@@ -511,12 +513,17 @@ int verb_status(int argc, char *argv[], void *userdata) {
                                 printf(" Net Boot URL: %s\n", loader_url);
 
+                        if (sysfail_entry)
+                                printf("System Failure Reason: %s\n", sysfail_reason);
```
the fields are aligned to the `:`. please stick to that. i.e. shorten System Failure to SysFail?

poettering commented on src/bootctl/bootctl.c:
```diff
@@ -236,6 +236,7 @@ static int help(int argc, char *argv[], void *userdata) {
                "\n%3$sBoot Loader Interface Commands:%4$s\n"
                "  set-default ID       Set default boot loader entry\n"
                "  set-oneshot ID       Set default boot loader entry, for next boot only\n"
+               "  set-sysfail ID       Set boot loader entry to be used for a system failure\n"
```
"for" → "in case of"

igoropaniuk replied:
Rephrased it,  changed "to be used"-> "used"

igoropaniuk commented:
@poettering Some CI issues have been reported (specifically, the `TEST-24-CRYPTSETUP` integration test is failing), but seems they are unrelated to this PR.

igoropaniuk commented:
@poettering @yuwata All comments have been addressed.

poettering commented on src/boot/boot.c:
```diff
@@ -2970,6 +3005,7 @@ static EFI_STATUS run(EFI_HANDLE image) {
         _cleanup_free_ char16_t *loaded_image_path = NULL;
         (void) device_path_to_str(loaded_image->FilePath, &loaded_image_path);
         config_load_all_entries(&config, loaded_image, loaded_image_path, root_dir);
+        sysfail_process(&config);
```
cast to void if you ignore the return value knowingly. alternatively make the func void

igoropaniuk replied:
fixed, thanks!

poettering commented:
lgtm (except for that nitpick). but needs a rebase

igoropaniuk commented:
@poettering Nitpick addressed, I've rebased on the latest main. Thanks!

igoropaniuk commented:
@poettering I've checked the logs for CI issues that have been reported, but it seems they are unrelated to this PR again.

YHNdnzj commented on man/systemd-boot-clear-sysfail.service.xml:
```diff
@@ -0,0 +1,52 @@
+  "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd">
+<!-- SPDX-License-Identifier: LGPL-2.1-or-later -->
+
+<refentry id="systemd-boot-clear-sysfail.service" conditional='ENABLE_BOOTLOADER'
```
The meson rules are not re-generated?

bluca commented on man/systemd-boot-clear-sysfail.service.xml:
```diff
@@ -0,0 +1,52 @@
+    'LoaderEntrySysFail' boot loader entry if the boot was successful and the 'LoaderSysFailReason' EFI variable,
+    which indicates the reason for the system failure, is not set. </para>
+
+    <para>The <filename>systemd-boot-random-seed.service</filename> unit invokes the <command>bootctl --graceful
```
copypasta
